Support 'template string' ('string interpolation')

aiya000 opened the pull request:
This motivation is here => #4491 :)

This is a version that omitted the '_' support.
(Please see https://github.com/vim/vim/issues/4491#issuecomment-509065896)
Please take this patch if you agree to this :+1:

Or let's discuss again at #4491 if Bram really want to '\_' :D

k-takata commented on runtime/doc/message.txt:
```diff
@@ -769,6 +769,7 @@ a user-defined command.
 This can only happen when changing the source code, when adding a command in
 src/ex_cmds.h.  The lookup table then needs to be updated, by running: >
 	make cmdidxs
+<
```
This should be removed.

aiya000 replied:
done :+1:

k-takata commented:
Please also check my comments in this commit: https://github.com/aiya000/vim/commit/e4e7d0da0126d5f90e00ef6b0148fdfc4608c1cb.

mattn commented on runtime/doc/eval.txt:
```diff
@@ -1331,6 +1331,57 @@ to be doubled.  These two commands are equivalent: >
+$. This is useful to avoid confusing a lot of repeated $.
+>
+	echo $'${10} ${20} ${30}'   " unreadable
+	echo _'${10} ${20} ${30}'   " readable!
```
should be removed.

aiya000 replied:
done :+1:

mattn commented on runtime/doc/eval.txt:
```diff
@@ -1331,6 +1331,57 @@ to be doubled.  These two commands are equivalent: >
+
+	" Only unscoped variables can omit {}
+	let x = 10
+	echo _'$x'  " 10
```
ditto

aiya000 replied:
I replaced to `$'x'` :+1:

mattn commented on runtime/doc/eval.txt:
```diff
@@ -1331,6 +1331,57 @@ to be doubled.  These two commands are equivalent: >
+	echo _'$x'  " 10
+
+	" In template-string, you can write '$' as $$
+	echo _'$$'  " $
```
ditto

aiya000 replied:
I replaced to `$'$$'`, thanks :D

mattn commented on runtime/doc/eval.txt:
```diff
@@ -1331,6 +1331,57 @@ to be doubled.  These two commands are equivalent: >
+	echo _'$$'  " $
+
+	" Escaping in template (non-literal) string
+	echo _"${10} \\ \""   " 10 \ "
```
ditto

aiya000 replied:
I replaced to
```vim
	echo $"${10} \\ \""   " 10 \ "
```

mattn commented on runtime/doc/eval.txt:
```diff
@@ -1331,6 +1331,57 @@ to be doubled.  These two commands are equivalent: >
+Exceptions:
+>
+	echo $'${var'  " E1000: Unterminated template literal: $'${var'
+	echo _'... $ ...'  " E1000: Missing variable name on a $: $'... $ ...'
```
ditto

aiya000 replied:
Thanks!
```vim
	echo $'... $ ...'  " E1000: Missing variable name on a $: $'... $ ...'
```

mattn commented on src/eval.c:
```diff
@@ -4569,6 +4613,39 @@ eval7(
 		break;
 
+    /*
+     * Template string constant: $"string ${var}", $'literal string ${var}', _"", _''
```
>Template string constant: $"string ${var}"

aiya000 replied:
```vim
     * Template string constant: $"string ${var}", $'literal string ${var}'
```
:+1:

mattn commented on src/eval.c:
```diff
@@ -5394,6 +5465,774 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+}
+
+/*
+ * Allocate a variable for a $'string ${var} $x', $"", _'', _"" constant.
```
>Allocate a variable for a $'string ${var} $x', $"" constant.

aiya000 replied:
:+1:
```
 * Allocate a variable for a $'string ${var} $x' constant.
```

mattn commented on src/eval.c:
```diff
@@ -5394,6 +5465,774 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+	char_u* template,
+	int evaluate)
+{
+    const size_t	QUOTES_LENGTH = 3;  // a number of strlen("$''"), strlen("$\"\""), strlen("_''"), or strlen("_\"\"")
```
>// a number of strlen("$''"), strlen("$\"\"")

aiya000 replied:
:+1:
```c
    const size_t	QUOTES_LENGTH = 3;  // a number of strlen("$''") (or strlen("$\"\""))
```

aiya000 commented:
@k-takata @mattn  Thanks for your reviews!
I fixed reviewd terms :+1:

And I re-checked comments. I think all comments are fixed :D

k-takata commented on runtime/doc/eval.txt:
```diff
@@ -1331,6 +1331,49 @@ to be doubled.  These two commands are equivalent: >
+
+Note that either single quotes or double quotes are used.
+
+These strings supports 'string-interpolations'. That embeds any expressions by
```
s/supports/support/

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,774 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    // NOTE: At here, "${}" of "${" "expr" "}", that called 'embedding', and "expr" called 'embedded'.
+
+    template = read_template_string(is_closing_quote, *arg);
+    if (template == NULL) {
```
This `{` should be the next line. (or should be removed, because this block is only one line.)

aiya000 replied:
:sunglasses:  :+1:
I did `/\v\) \{/` and `f s<CR>` all!

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,774 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    size_t		template_length;
+
+    for (p = source + OPENING_QUOTE_LENGTH; !(is_closing_quote(p) || *p == NUL); MB_PTR_ADV(p)) {}
+    if (*p == NUL) {
```
This `{` should be the next line.

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,774 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+	    mb_ptr_adv_nth(&to_dest, STRLEN(embedded[finished]->stringified) - 1);  // same
+	    ++finished;
+	    continue;
+	} else if (MB_PTR2LEN(to_source) > 1) {
```
`else` should be the next line.
`{` should be the next of the `else` line.

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,774 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    for (; !is_closing_quote(p); MB_PTR_ADV(p))
+    {
+	switch (*p) {
```
This `{` should be the next line.
Can we use `if (*p == NUL)`?

aiya000 replied:
!!
I fixed!

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,774 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    for (i = 0; i < SNEAK_CHARS_SIZE; ++i)
+    {
+	if (SNEAK_CHARS[i] == x) {
```
This `{` should be the next line.

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -203,6 +203,29 @@ static struct vimvar
+    /*
+     * "some_literal" or "var" or else.
+     */
+    char_u* expr;
```
```suggestion
    char_u *expr;
```

`*` should be prepended before a variable. Same for other declarations.

aiya000 replied:
Thanks a lot!
I did `:%s/\v([a-zA-Z_]+)\* ([a-zA-Z_])/\1 *\2/gc` :D

aiya000 commented:
[Appveyor](https://ci.appveyor.com/project/chrisbra/vim/builds/25868156/job/59t4uc2smotqjvox) says

```
gcc -c -I. -Iproto -DWIN32 -DWINVER=0x0600 -D_WIN32_WINNT=0x0600 -DHAVE_PATHDEF -DFEAT_NORMAL -DHAVE_STDINT_H -DHAVE_GETTEXT -DHAVE_LOCALE_H -DDYNAMIC_GETTEXT -DFEAT_CSCOPE -DFEAT_MBYTE_IME -DDYNAMIC_IME -DDYNAMIC_ICONV -pipe -march=i686 -Wall -O2 iscygpty.c -o obji686/iscygpty.o -U_WIN32_WINNT -D_WIN32_WINNT=0x0600 -DUSE_DYNFILEID -DENABLE_STUB_IMPL
gcc -I. -Iproto -DWIN32 -DWINVER=0x0600 -D_WIN32_WINNT=0x0600 -DHAVE_PATHDEF -DFEAT_NORMAL -DHAVE_STDINT_H -DHAVE_GETTEXT -DHAVE_LOCALE_H -DDYNAMIC_GETTEXT -DFEAT_CSCOPE -DFEAT_MBYTE_IME -DDYNAMIC_IME -DDYNAMIC_ICONV -pipe -march=i686 -Wall -O2 -s -municode -o vim.exe obji686/arabic.o obji686/autocmd.o obji686/beval.o obji686/blob.o obji686/blowfish.o obji686/buffer.o obji686/change.o obji686/charset.o obji686/crypt.o obji686/crypt_zip.o obji686/debugger.o obji686/dict.o obji686/diff.o obji686/digraph.o obji686/edit.o obji686/eval.o obji686/evalfunc.o obji686/ex_cmds.o obji686/ex_cmds2.o obji686/ex_docmd.o obji686/ex_eval.o obji686/ex_getln.o obji686/fileio.o obji686/findfile.o obji686/fold.o obji686/getchar.o obji686/hardcopy.o obji686/hashtab.o obji686/indent.o obji686/insexpand.o obji686/json.o obji686/list.o obji686/main.o obji686/mark.o obji686/memfile.o obji686/memline.o obji686/menu.o obji686/message.o obji686/misc1.o obji686/misc2.o obji686/move.o obji686/mbyte.o obji686/normal.o obji686/ops.o obji686/option.o obji686/os_mswin.o obji686/os_win32.o obji686/pathdef.o obji686/popupmnu.o obji686/popupwin.o obji686/quickfix.o obji686/regexp.o obji686/screen.o obji686/search.o obji686/sha256.o obji686/sign.o obji686/spell.o obji686/spellfile.o obji686/syntax.o obji686/tag.o obji686/term.o obji686/textprop.o obji686/ui.o obji686/undo.o obji686/usercmd.o obji686/userfunc.o obji686/version.o obji686/winclip.o obji686/window.o obji686/os_w32exe.o obji686/vimrc.o obji686/if_cscope.o obji686/xdiffi.o obji686/xemit.o obji686/xprepare.o obji686/xutils.o obji686/xhistogram.o obji686/xpatience.o obji686/iscygpty.o -lkernel32 -luser32 -lgdi32 -ladvapi32 -lcomdlg32 -lcomctl32 -lnetapi32 -lversion -lole32 -luuid      
obji686/popupwin.o:popupwin.c:(.text+0x2ee1): undefined reference to `ch_log'
collect2.exe: error: ld returned 1 exit status
mingw32-make.exe: *** [Make_cyg_ming.mak:1022: vim.exe] Error 1
```

But I don't have any ideas about this.
Is this my problem? X)

k-takata commented on src/eval.c:
```diff
@@ -6115,10 +6119,9 @@ read_next_string_length(char_u* whole_string, int expect_literal_string, char_u*
-	    case NUL:
-		semsg(_("E115: Missing quote: %s"), template);
-		return -1;
+	if (*p == NUL) {
```
`{` should be the next line.

aiya000 replied:
Oh my hands... lol
Thanks!

k-takata commented:
> But I don't have any ideas about this.
> Is this my problem? X)

Maybe no.
ref: https://groups.google.com/d/msg/vim_dev/zGy72qFwvYc/ZN9tZUpKBAAJ

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    const size_t	QUOTES_LENGTH = 3;  // a number of strlen("$''") (or strlen("$\"\""))
+    const size_t	OPNEING_QUOTE_LENGTH = 2;  // a number of strlen("$'")
+    size_t		i;
+    char_u*		result;
```
Could you move this `*` before the variable name?
(And there are similar cases remaining.)

aiya000 replied:
Sorry, I didn't replace it.

I'll `:%s/\v([a-zA-Z_]+)(\*+)([ 	])([\(a-zA-Z_])/\1\3\2\4/gc`!

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    size_t		i;
+    char_u*		result;
+    size_t		embedded_size;  // a size of embedded
+    embedded_t**	embedded = read_embedded(template, &embedded_size);
```
Same for this `**`.

aiya000 replied:
I'll `%s/\v([a-zA-Z_]+)(\*+)([ 	])([\(a-zA-Z_])/\1\3\2\4/gc` :+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+	memcpy(to_dest, fixed, fixed_size);
+	free(fixed);
+	mb_ptr_adv_nth(&to_dest, fixed_size - 1);  // 1 incresed by next for's effect
```
s/incresed/increased/

aiya000 replied:
I'll fix :+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    char_u*		template;
+    size_t		template_length;
+
+    for (p = source + OPENING_QUOTE_LENGTH; !(is_closing_quote(p) || *p == NUL); MB_PTR_ADV(p));
```
It might be better to move the last `;` to the next line (with indent) to emphasize that this is an empty loop.

aiya000 replied:
I'll do it :D

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    static embedded_t*
+new_embedded(int is_embedded_literal, char_u *expr, size_t expr_size)
+{
+    embedded_t *x = alloc(sizeof(embedded_t));
```
`ALLOC_ONE` can be used.
s/alloc(sizeof(embedded_t))/ALLOC_ONE(embedded_t)/

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    // Eval the result (non template) string
+    origin_string = string;
+    eval1(&string, rettv, evaluate);
+    free(origin_string);
```
`vim_free()` should be better than `free()` in Vim's source code.

aiya000 replied:
Great!
I'll `:%s/\<free\>/vim_free/gc` for my codes :sunglasses:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+	{
+	    ++count;
+	    ++p;      // a part of "$"
+	    continue; // a part of "{" (incresing)
```
s/incresing/increasing/

aiya000 replied:
I'll fix!

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+ */
+    static int
+get_template_string_tv(
+	char_u* (*enclose)(char_u*),
```
s/char_u* (/char_u *(/

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+/*
+ * Get a first "expr" of "${" "expr" "}" from `whole_embedding`.
+ *
+ * whole_embedding starts with "${": e.g. "${10}`", "${var} Hello :D`".
```
The two backquotes can be removed?

aiya000 replied:
I'll `:%s/MUST BE `free`d/MUST BE EXECUTED BY free()/gc`, and I'll search via ``/\v`[a-zA-Z_]``` and remove enclosing ``!

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+/*
+ * Get a first "var" of "$" "var" from `whole_embedding`.
+ *
+ * whole_embedding starts with "$": e.g. "$var`", "$v_i_m".
```
The backquote can be removed?

aiya000 replied:
:+1:

k-takata replied:
I meant that ```"$var`"``` isn't a mistake of ```"$var"```?
I didn't intend that you should remove all backquotes from the comments.

aiya000 replied:
Sorry, my reply was too meaningless x(
`` "$var` `` is not intended, but `"$var'"` is intended.

I replaced to
```
 * whole_embedding starts with "$": e.g. "$var", "$var'", "$v_i_m".
```
:+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    for (p = expr_start + 1; *p != NUL; MB_PTR_ADV(p))
+    {
+	if (!is_sneak_case_char(*p))
```
Does this mean that numbers cannot be used for a variable name?

aiya000 replied:
No! I don't intend x)

I'll add tests, And fix this!

k-takata commented on src/testdir/Make_all.mak:
```diff
@@ -283,7 +283,8 @@ NEW_TESTS = \
 	test_alot_utf8 \
-	test_alot
+	test_alot \
+	test_string_interpolation
```
This list should be in the alphabetical order (except the last `test_alot` tests).

aiya000 replied:
I'll fix this :+1:

aiya000 commented:
@k-takata Thanks a lot!
I might resolve problems :+1:

k-takata commented on src/eval.c:
```diff
@@ -244,6 +267,28 @@ static int eval7(char_u **arg, typval_T *rettv, int evaluate, int want_string);
 static int get_string_tv(char_u **arg, typval_T *rettv, int evaluate);
 static int get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate);
+static int get_template_string_tv(char_u *(*enclose)(char_u*), char_u *(*disclose)(char_u*), char_u *(*fix_quote)(char_u), int (*is_closing_quote)(char_u*), char_u **arg, typval_T *rettv, int evaluate);
+static char_u* disclose_quote(char_u *quoted);
```
Please adjust the place of `*`.

aiya000 replied:
Sorry x(
What do you mean?

`static char_u *disclose_quote(char_u *quoted);` ?

k-takata replied:
yes

aiya000 replied:
Thanks!
I fixed!

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5466,805 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+{
+    const size_t	QUOTES_LENGTH = 2;  // strlen("''")
+    size_t		result_length = STRLEN(quoted) - QUOTES_LENGTH;
+    char_u*		result;
```
ditto

aiya000 replied:
Oh x(
Fixed :+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5466,805 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+read_template_string(int (*is_closing_quote)(char_u*), char_u *source)
+{
+    const size_t 	OPENING_QUOTE_LENGTH = 2;  // "$'"
+    char_u*		p;
```
ditto

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5466,805 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    for (; *to_source != NUL; MB_PTR_ADV(to_dest), MB_PTR_ADV(to_source))
+    {
+	char_u*		fixed = fix_quote(*to_source);
```
ditto

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5466,805 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+{
+    const size_t	LITERAL_EMBEDDING_START_LENGTH = 2;  // a number of "${"
+    const size_t	VAR_EMBEDDING_START_LENGTH = 1;  // a number of head "$"
+    char_u*		to_dest;
```
ditto

aiya000 replied:
wa!?
Sorry, I failed replacing...

I did `:%s/\v([a-zA-Z_]+)(\*+)([ 	]+)/\1\3\2/gc` again!

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5466,805 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    const size_t	VAR_EMBEDDING_START_LENGTH = 1;  // a number of "$"
+    embedded_t	**embedded;
+    size_t		finished = 0;
+    char_u*		p;
```
ditto

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5466,805 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+read_next_embedded_lit(char_u *whole_embedding, char_u *template)
+{
+    const size_t	LITERAL_EMBEDDING_START_LENGTH = 2;  // "${"
+    char_u*		embedded_start = whole_embedding + LITERAL_EMBEDDING_START_LENGTH;
```
ditto

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5466,805 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+read_next_embedded_var(char_u *whole_embedding, char_u *template)
+{
+    const size_t	NUM_LAST_NON_SNEAK_CHAR = 1;
+    char_u*		expr_start = whole_embedding + 1;  // skip a head '$'
```
ditto

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -6130,11 +6132,11 @@ read_next_string_length(char_u *whole_string, int expect_literal_string, char_u
+ * Get a first "expr" of "${" "expr" "}" from whole_embedding.
  *
- * whole_embedding starts with "${": e.g. "${10}`", "${var} Hello :D`".
+ * whole_embedding starts with "${": e.g. "${10}'", "${var} Hello :D'".
```
Are the changes of this line really correct?
Shouldn't be `"${10}"` and `"${var} Hello :D"`?

aiya000 replied:
Yes, I intended this :D
e.g. `$'hello ${var} and ${10}'` passes `"${var} and ${10}'` and `"${10}'` to this function!

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5466,805 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    char_u*		expr_start = whole_embedding + 1;  // skip a head '$'
+    char_u*		p;
+
+    if (!is_sneak_case_head_char(expr_start[0]))
```
I wonder if we really need this function.
Isn't `!ASCII_ISALPHA(expr_start[0]) && expr_start[0] != '_'` enough?

aiya000 replied:
In my opinion, it allows `"0aaa"` and else (cases starts with `[0-9]`), 
    and doesn't allow `_name` (cases starts with `_`).
    ( Sorry if this is wrong x( )
But I want to allow only cases of `[a-zA-Z_][a-zA-Z0-9_]+`.

:smile:

aiya000 replied:
Sorry, I understood. this is not needed! I was perfectly wrong ;D
I'll use ASCII_ISALPHA instead of this function!

k-takata commented on src/eval.c:
```diff
@@ -5394,6 +5466,805 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    for (p = expr_start + 1; *p != NUL; MB_PTR_ADV(p))
+    {
+	if (!is_sneak_case_char(*p))
```
I wonder if we really need this function.
Isn't `!ASCII_ISALNUM(*p) && *p != '_'` enough?

aiya000 replied:
Exactly xD
I fixed!

aiya000 commented:
@k-takata  Thanks :+1:
I might fix again :D

k-takata commented:
According to the result of codecov, it seems that your test is not executed on CI.
You also need to add

```vim
source test_string_interpolation.vim
```

in `src/testdir/test_alot.vim` (or add `test_string_interpolation.res` in the `NEW_TESTS_RES` variable in `src/testdir/Make_all.mak`).

k-takata commented on src/testdir/test_string_interpolation.vim:
```diff
@@ -0,0 +1,123 @@
+  call s:illformed()
+endfunc
+
+func s:basic() abort
```
It might be better to make this function global. E.g. `Test_string_interpolation_basic`.
If it is global, you can run a specific test. See:
https://github.com/vim/vim/blob/e28cfb28126c07abd9e9ba04e36b3789433201b6/src/testdir/runtest.vim#L5-L7

aiya000 replied:
I'll add it :+1:

k-takata commented on src/testdir/test_string_interpolation.vim:
```diff
@@ -0,0 +1,123 @@
+  let x = $'${10}'
+endfunc
+
+func s:appendix() abort
```
ditto

aiya000 replied:
:+1:

k-takata commented on src/testdir/test_string_interpolation.vim:
```diff
@@ -0,0 +1,123 @@
+endfunc
+
+" These should be an exception, should not be a 'Segmentation fault'.
+func s:illformed() abort
```
ditto

aiya000 replied:
:+1:

k-takata commented on src/testdir/test_string_interpolation.vim:
```diff
@@ -0,0 +1,123 @@
+
+scriptencoding utf-8
+
+func Test_string_interpolation()
```
This function is not needed, if you make all the static functions global.

aiya000 replied:
:+1:

aiya000 commented:
@k-takata  Sorry, Now I'm working to fix problems.
Please wait for re-checking :pray:

aiya000 commented:
Sorry, accidentally I closed this x(
I reopened!

aiya000 commented:
> According to the result of codecov, it seems that your test is not executed on CI.
> You also need to add
> 
> source test_string_interpolation.vim
> in src/testdir/test_alot.vim (or add test_string_interpolation.res in the NEW_TESTS_RES variable in src/testdir/Make_all.mak).

Thanks! I added! (as test_template_string.vim. please see below :pray:)

- - -

I made these commits!

- 4e17dd3f4 2019-07-12 21:58 Rename test_string_interpolation to test_template_string
- 783f9769a 2019-07-12 21:58 Add missing test_template_string to test_alot.vim
- 12b25b776 2019-07-12 22:43 Refine test_template_string
- c5c709eae 2019-07-14 16:13 Add a test to empty template literals
- 78a4d2d79 2019-07-14 16:13 Add a help for empty template literals
- 1e7a03cf8 2019-07-14 18:33 Add a test
- 8b81192d1 2019-07-14 18:33 Fix a test of empty template literals
- 0706cb5a2 2019-07-14 22:21 Remove abort from tests
- 27efec2e2 2019-07-14 23:00 Resolve tsuyoshi_cho's review

For detail:

- 4e17dd3f4 2019-07-12 21:58 Rename test_string_interpolation to test_template_string
- 783f9769a 2019-07-12 21:58 Add missing test_template_string to test_alot.vim
- 12b25b776 2019-07-12 22:43 Refine test_template_string

:point_up:
These renames `test_string_interpolation` to `test_template_string`,
    then adds it to Make_all.mak.

- c5c709eae 2019-07-14 16:13 Add a test to empty template literals
- 78a4d2d79 2019-07-14 16:13 Add a help for empty template literals
- 8b81192d1 2019-07-14 18:33 Fix a test of empty template literals

:point_up:
mattn tell me a sensitive case (`$'${}'`).
This adds it as a test, then fixes it.
Also adds a help.

- 1e7a03cf8 2019-07-14 18:33 Add a test

:point_up: Adds a minimum test

```
+ call assert_equal('10 and me', $'$x and me')
```

- 0706cb5a2 2019-07-14 22:21 Remove abort from tests

:point_up: Suddenly, I met two test failing after `test_string_interpolation` renamed to `test_template_string`.

This first is :point_down:

```
From test_template_string.vim:
Executed Test_template_string_appendix() in   0.000358 seconds
Executed Test_template_string_basic() in   0.000271 seconds
Executed Test_template_string_illformed() in   0.000232 seconds
Executed abort in   0.000279 seconds
Executed abort in   0.000279 seconds
Executed abort in   0.000328 seconds
Executed 6 tests in   1.008782 seconds
3 FAILED:
Found errors in abort:
Caught exception in abort: Vim(call):E107: Missing parentheses: abort @ function RunTheTest, line 40
Found errors in abort:
Caught exception in abort: Vim(call):E107: Missing parentheses: abort @ function RunTheTest, line 40
Found errors in abort:
Caught exception in abort: Vim(call):E107: Missing parentheses: abort @ function RunTheTest, line 40
make: *** [Makefile:78: test_template_string] Error 1
```

This commit fixes it.

- 27efec2e2 2019-07-14 23:00 Resolve tsuyoshi_cho's review

:point_up: The rest one is :point_down:

```
Found errors in Test_template_string_basic():
function RunTheTest[40]..Test_template_string_basic line 3: Expected 'function(''function'')' but got 'function(''function'')ct\x01\x01'
```

This commit fixes it!

- - - - -

Thanks :smiley:

- - - - -

@k-takata 
Sorry for your waiting.
I might fix mentioned terms!

Please check again :bow:

aiya000 commented:
wa!?  test failure!?

aiya000 commented:
@k-takata Hi, thanks for your waiting :D

I did

1. Pass tests for CI
1. Fix memory leaks via valgrind
    - Checked with `VALGRIND = valgrind --tool=memcheck --leak-check=yes --num-callers=25 --log-file=valgrind.$* --track-origins=yes`
        - at [`src/testdir/Makefile:L21`](https://github.com/vim/vim/blob/df9c6cad8cc318e26e99c3b055f0788e7d6582de/src/testdir/Makefile#L21)
    - [current status](https://gist.github.com/a023090e2c9c0447193154915aae371c)

The 'current status' has a lot of 'possibly lost'.
But in my opinion, these are not my `get_template_string_tv()`'s memory leaks.
...Is this right?

I might completely fix problems!

I'll `git-squash` all commits if it is truth and if needed.

Thanks :smiley:

k-takata commented on src/eval.c:
```diff
@@ -244,6 +267,28 @@ static int eval7(char_u **arg, typval_T *rettv, int evaluate, int want_string);
+static int is_closing_double_quote(char_u *x);
+static char_u *duplicate_single_quote(char_u x);
+static char_u *dont_fix(char_u x);
+static char_u* parse_template(char_u *(*enclose)(char_u*), char_u *(*disclose)(char_u*), int (*should_be_escaped)(char_u), char_u *(*fix_quote)(char_u), int (*is_closing_quote)(char_u*), char_u *template, int evaluate);
```
`s/char_u* /char */`

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -5407,6 +5481,879 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+ * also set the last and all to NUL.
+ */
+    static char_u*
+alloc_terminated(size_t size)
```
Is this function really needed?
Isn't this the same as `(char_u*)alloc_clear(size + 1)`?

aiya000 replied:
I'll remove it!

k-takata commented on src/testdir/test_template_string.vim:
```diff
@@ -0,0 +1,129 @@
+  try
+    let _ = $"missing closing double quote
+    call assert_report('Should throw an exception.')
+  catch /^Vim(let):E115/
```
`s/E115/E115:`

aiya000 replied:
:+1:

k-takata commented on src/testdir/test_template_string.vim:
```diff
@@ -0,0 +1,129 @@
+    " syntax error in a embedding
+    let _ = $'missing closing of ${function(}'
+    call assert_report('Should throw an exception.')
+  catch /^Vim(let):E119/
```
`s/E119/E119:`

aiya000 replied:
:+1:

k-takata commented on src/testdir/test_template_string.vim:
```diff
@@ -0,0 +1,129 @@
+  try
+    let _ = $'missing closing quote
+    call assert_report('Should throw an exception.')
+  catch /^Vim(let):E115/
```
`s/E115/E115:`

k-takata commented on src/testdir/test_template_string.vim:
```diff
@@ -0,0 +1,129 @@
+  try
+    let _ = $'$10'
+    call assert_report('Should throw an exception.')
+  catch /^Vim(let):E1000: Missing variable name/
```
Error messages can be translated. This will fail when the language is not English.

aiya000 replied:
Good! I'll remove messages!

aiya000 commented:
@k-takata  Thank you as always :joy: 
I might apply all review!

aiya000 commented:
I did `git-squash` into only 3-commits :)

aiya000 commented:
I love a number 1000.
Please don't change `E1000` to any another number if Bram is ok :pray:

But this is a my indulgence, If Bram doesn't agree this, I agree changing :sob:

brammool commented:
I'm afraid that this is too much code for functionality that can also be done with string concatenation.  Without adding $"" plugin authors can do the same thing, perhaps not that nice, but it's not worth so much code for something that is just a bit nicer syntax.

aiya000 commented:
In my opinion, this is useful because `$""` style strings is easier to read than concat style `..` strings :)

This means (I think) `$""` seems nearly to the result text!

aiya000 commented:
@brammool
sorry, I misunderstood your message x(

How much do you expect code volume that you suppose?
(I'll do my best for it👍️)

brammool commented:
One reason to use a template is when translations are used.  Then the order of the expressions could be changed.  We don't have Vim script translation support, but it can be done in other ways (provide a dictionary with translations).  So let's see if we can make this patch simpler.

get_template_string_tv() takes a lot of arguments that are functions, I don't think that is good.  Especially because the functions have a lot of duplication.
I think you can use get_lit_string_tv() and get_string_tv() first, and then deal with the ${} items.
And use a growarray to build the result (garray_T) to avoid having to compute the size first.
I would think after detecting ${ you can call eval1(), then check the next non-white is }.

aiya000 commented:
Thanks!
I'll do it :+1:

I read this PR's algorithm again, also think this is bad x(
(Who is this writing!?)

> get_template_string_tv() takes a lot of arguments that are functions, I don't think that is good. Especially because the functions have a lot of duplication.
> I think you can use get_lit_string_tv() and get_string_tv() first, and then deal with the ${} items.

Indeed :+1:

> One reason to use a template is when translations are used. Then the order of the expressions could be changed. We don't have Vim script translation support, but it can be done in other ways (provide a dictionary with translations). So let's see if we can make this patch simpler.

This means `embedded_T`?

> And use a growarray to build the result (garray_T) to avoid having to compute the size first.
> I would think after detecting ${ you can call eval1(), then check the next non-white is }.

!!
Thanks. I'll read it :D

- - -

My new algorithm is here, this maybe little.

1. Read `$'${foo} ${10}'` from head
1. Get the result `'' .. {x -> (type(x) is v:t_string) ? x : string(x)}(foo) .. ' ' .. {x -> (type(x) is v:t_string) ? x : string(x)}(10)`
    - using `garray_T`
1. Pass it to `eval1()`
    - do `if` by `''` or `""` in the internal of `get_template_string_tv()` (not on the arguments of `get_template_string_tv()`)

Thanks a lot of advices :D

aiya000 committed:
Make a test for template string

aiya000 committed:
Add a document for template string

aiya000 committed:
Implement template string

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3617,293 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
     return OK;
 }
 
+int get_repeated_single_quotes_length(char_u *x, char_u *whole_start) {
```
Wrong coding style. This should be:

```c
    int
get_repeated_single_quotes_length(char_u *x, char_u *whole_start)
{
```

Cannot this function be static?

aiya000 replied:
This can be static :+1:

I fixed this to below

```c
    static int
get_repeated_single_quotes_length(char_u *x, char_u *whole_start)
{
    char_u *follower;
    char_u *following;
    ...
```

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3617,293 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    return (x - follower) + (following - x) + 1;  // 1 is num of *x
+}
+
+size_t is_closing_quote(char_u *template, int is_literal_string, char_u *whole_start)
```
Wrong coding style. `size_t` should be moved to the previous line.
Cannot this function be static?

aiya000 replied:
x)

I fixed

```c
    static size_t
is_closing_quote(char_u *template, int is_literal_string, char_u *whole_start)
{
```

k-takata commented on src/macros.h:
```diff
@@ -233,6 +233,8 @@
 /* Advance multi-byte pointer, skip over composing chars. */
 #define MB_PTR_ADV(p)	    p += has_mbyte ? (*mb_ptr2len)(p) : 1
+/* Decrease p with a character, like MB_PTR_ADV(). */
+#define MB_PTR_DECR(p)	    p -= has_mbyte ? (*mb_ptr2len)(p) : 1
```
This macro looks incorrect.
If `p` points the top of a two-byte character, this always decrements two bytes regardless of the length of the previous character.

I think that `MB_PTR_BACK()` should be used instead.

aiya000 replied:
Oh, I was wrong.

I fixed this as below because I think this checks only one byte characters.

```c
    static int
get_repeated_single_quotes_length(char_u *x, char_u *whole_start)
{
    char_u *follower;
    char_u *following;

    if (*x != '\'')
    {
        return 0;
    }

    for (follower = x; *follower == '\''; --follower)
    {
    }
    for (following = x; *following == '\''; ++following)
    {
    }

    return (x - follower) + (following - x) + 1;  // 1 is num of *x
}
```

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3617,293 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    for (follower = x; *follower == '\''; MB_PTR_DECR(follower))
+    {
+    }
+    for (following = x; *following == '\''; MB_PTR_ADV(following))
```
If `following` points `'`, it is obviously a single byte character, so no need to use `MB_PTR_ADV` to advance the pointer. (Using `MB_PTR_ADV` might be harmless, though.)

aiya000 replied:
Indeed, thanks!
I'll do it :+1:

aiya000 commented:
@k-takata  Thanks for your reviews!
I might fix all reviews :) (https://github.com/vim/vim/pull/4634/commits/8590de884ce32747b288682e6db843dd50ed1ba7)

I'll continue to fix memory leaks.
Please tell problems to me if it is existent x)

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3617,296 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+        return 0;
+    }
+
+    for (follower = x; *follower == '\''; --follower)
```
If there is a multibyte encoding which trailing byte can be `'`, this might not work well.
I'm not sure whether such strange encoding really exists. (At least, this is safe for UTF-8, Shift_JIS and EUC-JP.)

aiya000 replied:
I think so too.
mmm... Is a counterplan for this needed?

aiya000 replied:
@k-takata I removed this algorithm from the implementation!
Please see new implementation :pray:

aiya000 commented:
I killed all `definiately losts` of valgrand!
Also I did `git-squash` into 3 commits.

@k-takata Please review latest implementation :pray:

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3617,304 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+}
+
+    static int
+is_closing_quote(char_u *template, int is_literal_string, int *is_skipping_needed)
```
`is_closing_double_quote()` and `is_closing_single_quote()` do almost the same thing.
I think they can be merged into this function.

aiya000 replied:
mmm...
Indeed, this leaves one more optimization.
But I want to split the function to out of `is_closing_quote()`.

I arranged as below :+1:

```c
    static int
is_expected_closing(char_u expected, char_u exclude1, char_u exclude2, char_u *template, int *is_skipping_needed)
{
    if (*template == exclude1 && *(template + 1) == exclude2)
    {
	*is_skipping_needed = TRUE;
        return FALSE;
    }

    *is_skipping_needed = FALSE;
    return *template == expected;
}

    static int
is_closing_quote(char_u *template, int is_literal_string, int *is_skipping_needed)
{
    return is_literal_string
	? is_expected_closing('\'', '\'', '\'', template, is_skipping_needed)
	: is_expected_closing('"', '\\', '"', template, is_skipping_needed);
}
```

k-takata replied:
I would merge `is_expected_closing()` into `is_closing_quote()` and use `is_escaped_quote()` for the first `if` condition.

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3617,304 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    if (!ASCII_ISALPHA(**expr) && **expr != '_')
+    {
+	vim_free(result.ga_data);
```
`ga_clear(&result)` might be better?

aiya000 replied:
Thanks :+1:

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3617,304 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    {
+	if (**expr == NUL)
+	{
+	    vim_free(result.ga_data);
```
`ga_clear(&result)` might be better?

aiya000 replied:
Thanks!
I'll fix :+1:

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3617,304 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+		{
+		    is_in_quote = !is_in_quote;
+		}
+	    }
```
Outer pair of { } looks unneeded.

It might be better to add `break;`.

aiya000 replied:
I think so too, thanks! :+1:

aiya000 commented:
Why coverage/covdralls fail? x(

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3617,304 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    {
+	int is_escaped_quote =
+	    (is_literal_string && *p == '\'' && *(p + 1) == '\'') ||
+	    (!is_literal_string && *p == '\\' && *(p + 1) == '"');
```
These kind of checks are appeared several times. How about making this as a function?
E.g. `is_escaped_quote(is_literal_string, p)`

aiya000 replied:
:+1:

I added the declaration,

```c
static int is_escaped_quote(int is_literal_string, char_u *quotes);
```

defined this,

```c
    static int
is_escaped_quote(int is_literal_string, char_u *p)
{
    return
	(is_literal_string && *p == '\'' && *(p + 1) == '\'') ||
	(!is_literal_string && *p == '\\' && *(p + 1) == '"');
}
```

also used it.

```c
	if (is_escaped_quote(is_literal_string, p))
	{
	    ga_append(&result, *(p + 1));
	    MB_PTR_ADV(p);
	    continue;
	}
	ga_append(&result, *p);
```

k-takata commented on src/testdir/test_template_string.vim:
```diff
@@ -0,0 +1,132 @@
+  " Another types (compound tests)
+  if exists('*job_start')
+    let x = job_start('ls', {})
+    call assert_equal(string(x), $'${x}')  " About the variable expansion is belo, $'${x}'w
```
Typo in comment. "below" is separated.

aiya000 replied:
oops x)
Thanks.

I remove this commment, because it seems unnecessary.

k-takata commented on src/testdir/test_template_string.vim:
```diff
@@ -0,0 +1,132 @@
+    let _ = $'missing closing quote
+    call assert_report('Should throw an exception.')
+  catch /^Vim(let):E115:/
+    " Good
```
It might be better to use `assert_exception()`?

aiya000 replied:
:+1:
I replaced

```vim
  try
    let _ = $'missing closing quote
    call assert_report('Should throw an exception.')
  catch
    call assert_exception('E115:')
  endtry
```

k-takata commented on runtime/doc/eval.txt:
```diff
@@ -1378,6 +1378,51 @@ to be doubled.  These two commands are equivalent: >
+	echo $'${var'  " E1000: Unterminated template literal: $'${var'
+	echo $'$10'    " E1000: Illegal template literal: $'$10'
+	echo $'... $ ...'  " E1000: Missing variable name on a $: $'... $ ...'
+	echo $'---${}---'  " E1000: Empty template literal: $'---${}---'
```
I think that different error messages should use different error number.

aiya000 replied:
mmm, I want to use `E1000`.
Because Now `E999` is the maximum number of exceptions, and I love a number `1000` x(

But I cannot refuse this if this is really needed x)

k-takata replied:
Just FYI, there are still many error numbers available before E1000.
https://github.com/vim/vim/blob/e8a7dfedfc8ea5c376c7912cb27a9405e4b8c972/runtime/doc/todo.txt#L132-L135
E.g. E450 to E453 might be able to be used for these 4 error messages.

aiya000 replied:
Understood, I'll use these number before E1000 😎

Now I'm using only E1000 for new behaviors.
Should I use different numbers (e.g. E450 to E453) for these?

k-takata replied:
Bram will decide the error numbers when merging. You can use arbitrary numbers while developing.

> Should I use different numbers (e.g. E450 to E453) for these?

Not sure.
At least, I think it's better to use different numbers. This has a merit that you can catch the different errors in your tests.

aiya000 replied:
Sorry, I understood.
I entrust about E1000 to Bram.
Thanks a lot!

Also the merit seems good.
I'll use these different numbers 👍️

aiya000 replied:
I fixed :+1:

```vim
  try
    let _ = $'$10'
    call assert_report('Should throw an exception.')
  catch
    call assert_exception('E450:')
  endtry

  try
    let _ = $'---$---'
    call assert_report('Should throw an exception.')
  catch
    call assert_exception('E450:')
  endtry

  try
    let _ = $'${10'
    call assert_report('Should throw an exception.')
  catch
    call assert_exception('E451:')
  endtry

  try
    let _ = $'---${}---'
    call assert_report('Should throw an exception.')
  catch
    call assert_exception('E452:')
  endtry
```

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3617,304 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+/*
+ * Parses a string like "${var}", and consume it onto "}".
+ *
+ * Return "{x->(type(x) is v:t_string?x:string(x)}(var)", or NULL if expr is invalid.
```
s/(type(x)/type(x)/

I think it is better to insert spaces for readability.
`{x -> type(x) is v:t_string ? x : string(x)}(var)`

aiya000 replied:
Sorry, this is an oloder comment.
The latest implementation returns `"{x -> type(x) is v:t_string ? x : string(x)}(var)"`.
I fixed this :pray:

```c
 * Return "{x -> type(x) is v:t_string ? x : string(x)}(var)", or NULL if expr is invalid.
```

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3617,304 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+/*
+ * Parses a string like "$var", and consume expr to the last char of a variable name.
+ *
+ * Return "{x->(type(x) is v:t_string?x:string(x)}(var)", or NULL if expr is invalid.
```
s/(type(x)/type(x)/

I think it is better to insert spaces for readability.
`{x -> type(x) is v:t_string ? x : string(x)}(var)`

aiya000 replied:
Thanks, I replaced the comment.

aiya000 commented:
@k-takata  Thanks a lot :D
I might apply all reviews :+1:

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3618,294 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    // an escaped single quote in the single quoted string,
+    int did_encount_single_single_in_single =
+	(is_literal_string && **expr == '\'' && *(*expr + 1) == '\'');
```
I would use `is_escaped_quote()` here.

aiya000 replied:
I'll use it :+1:

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3618,294 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    // or a single quote in the double quoted string?
+    int did_encount_escaped_double_in_double =
+	(!is_literal_string && **expr == '\\' && *(*expr + 1) == '"');
```
I would use `is_escaped_quote()` also here. Then this line can be merged with above `did_encount_single_single_in_single` line.

E.g.
```C
int did_encount_escaped_quote = is_escaped_quote(is_literal_string, *expr);
```

aiya000 replied:
I merged the two var into `did_encount_escaped_quote` :+1:

```c
    static int
did_encount_string_quote(char_u **expr, int is_literal_string)
{
    // encounted a double quote in the single quoted string (lit str),
    int did_encount_double_in_single =
	(is_literal_string && **expr == '"');

    // a single quote in the double quoted string,
    int did_encount_single_in_double =
	(!is_literal_string && **expr == '\'');

    // or a single quote in the double quoted string?
    int did_encount_escaped_quote = is_escaped_quote(is_literal_string, *expr);

    int result =
	did_encount_double_in_single ||
	did_encount_single_in_double ||
	did_encount_escaped_quote;

    if (did_encount_escaped_quote)
    {
	++*expr;
    }

    return result;
}
```

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3618,294 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+}
+
+    static int
+is_closing_quote(char_u *template, int is_literal_string, int *is_skipping_needed)
```
I intended something like this (Not tested):

```C
    static int
is_closing_quote(char_u *template, int is_literal_string, int *is_skipping_needed)
{
    char_u  quote = is_literal_string ? '\'' : '"';

    if (is_escaped_quote(is_literal_string, template))
    {
	*is_skipping_needed = TRUE;
	return FALSE;
    }

    *is_skipping_needed = FALSE;
    return *template == quote;
}
```

Isn't this simple enough even `is_expected_closing()` is merged into this?

aiya000 replied:
Thanks, I'll use it!

aiya000 commented:
@k-takata  Thanks again!
I might fixed all reviews.

k-takata commented on src/eval.c:
```diff
@@ -3605,6 +3618,282 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    garray_T	plain;  // a string that is not a template string
+    char_u	*orig_arg = *arg;
+    int		is_literal_string = (*(*arg + 1) == '\'');
+    char_u	quote = is_literal_string ? '\'' : '"';
```
How about this?
```C
    char_u	quote = *(*arg + 1);
    int		is_literal_string = (quote == '\'');
```
A bit simpler, I think.

aiya000 replied:
Good!
I'll fix it.

aiya000 commented:
@k-takata  I replaced the implementation around :+1:

- `char_u *stringified = enclose_template_literal_by_stringify(` of `get_template_string_tv()`
- `enclose_template_literal_by_stringify()`

The code size seems more decreased!

aiya000 commented:
In my opinion, this can be merged😎
Please do anything you like!
(Also sorry, please forget about `E1000`. Please put any number :D)

brammool commented:
Finally had time to look at this again.
It looks like $ is used as a special character.  Thus to get a literaly $ it needs to be doubled.  I don't see a good reason for this, and typescript also doesn't appear to do this.
Since encountering ${ in a string would be rare, I don't think escaping is needed.  Instead one would use $"text${'${'}text"

I don't like the use of "stringify_head".  Generating a lambda and then later interpreting it is quite inefficient.  Why not call eval1() on the text after ${, applying string() on it if needed and then appending the result to the growarray?  eval1() also skips over the expression.

For the text that is a string, we could just call get_string_tv(), or use the inner part only (to handle escaped characters).

aiya000 commented:
> It looks like $ is used as a special character. Thus to get a literaly $ it needs to be doubled. I don't see a good reason for this, and typescript also doesn't appear to do this.
> Since encountering ${ in a string would be rare, I don't think escaping is needed. Instead one would use
> $"text${'${'}text"

Do you mean `$"text${'$'}text"`?

I agree :+1:
I dropped the implementation.

> I don't like the use of "stringify_head". Generating a lambda and then later interpreting it is quite inefficient. Why not call eval1() on the text after ${, applying string() on it if needed and then appending the result to the growarray? eval1() also skips over the expression.

I'll do it.
Please wait a little term :)

> For the text that is a string, we could just call get_string_tv(), or use the inner part only (to handle escaped characters).

It is meaning instead of the last `eval1()`. isn't it?
I'll do it :+1:

brammool commented:
> > It looks like $ is used as a special character. Thus to get a
> > literaly $ it needs to be doubled. I don't see a good reason for
> > this, and typescript also doesn't appear to do this.
> > Since encountering ${ in a string would be rare, I don't think
> > escaping is needed. Instead one would use
> > $"text${'${'}text"
> 
> Do you mean `$"text${'$'}text"`?

Using "$" by itself is just a normal character, only "${" starts the
expression.  Thus only when you want "${" literally you need to use
"${'${'}".  I don't think there is a better way to escape this.  I could
not find what Typescript is using, I suspect it's the same.

> I agree :+1:
> I dropped the implementation.
> 
> > I don't like the use of "stringify_head". Generating a lambda and
> > then later interpreting it is quite inefficient. Why not call
> > eval1() on the text after ${, applying string() on it if needed and
> > then appending the result to the growarray? eval1() also skips over
> > the expression.
> 
> I'll do it.
> Please wait :)
> 
> > For the text that is a string, we could just call get_string_tv(), or use the inner part only (to handle escaped characters).
> 
> It is meaning instead of the last `eval1()`. isn't it?
> I'll do it :+1:

Good luck implementing this.

-- 
In a world without walls and borders, who needs windows and gates?

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

aiya000 commented:
Thank you for waiting!

I removed escaping behavior of `$'$$'`, implmenet to evaluate template literals successively using `garray_T`, and

> > For the text that is a string, we could just call get_string_tv(), or use the inner part only (to handle escaped characters).
> 
> It is meaning instead of the last eval1(). isn't it?
> I'll do it +1

In `get_template_string_tv()`, unfortunately , the `eval1()` couldn't be replaced.
Because the target string is not determined as either a literal string or a string :pray:

Thanks :D

aiya000 commented:
mmm... is the test results...strange?

k-takata commented:
> mmm... is the test results...strange?

Double-free occurred: https://travis-ci.org/vim/vim/jobs/629606342#L3023

aiya000 commented:
@k-takata thanks!
I fixed it :D

aiya000 commented:
In my opinion, all needed operations are done!
Please tell me if some operation is missing :)

aiya000 commented:
`git-rebase`d!

k-takata commented:
@aiya000 todo.txt has been updated.
https://github.com/vim/vim/blob/eed3571fe0f7d2ae779e5d36388cc4fe16c3516e/runtime/doc/todo.txt#L186-L187

Edit: misunderstood. The paragraph was not added, just moved above.

aiya000 commented:
@k-takata Thanks!
Should I implement without using lambda?

k-takata committed:
Fix coding style

k-takata committed:
Fix multibyte handling

k-takata committed:
Fix some issues

k-takata committed:
Fix misspelling

k-takata committed:
Fix handling of backslash

```
let x = '\x31'
echo $"${x}"
```

This should print "\x31" not "1".

k-takata commented:
I found some issues, so I opened a PR on your repository: https://github.com/aiya000/vim/pull/1.

mattn commented:
I prefer that this template strings is disabled in modeline.

mattn commented:
And also this should work only when `set sandbox` ?

aiya000 commented:
Thanks!

Is it meaning that `get_template_string_tv()` (the evaluation function for template strings) should return `FAIL` if `!&sandbox`?

aiya000 commented:
I'm thinking to change the spec to

```vim
  " Quote in quote
  call assert_equal('', $'${''}')
  call assert_equal('x', $'${'x'}')
  call assert_equal("", $"${""}")
  call assert_equal("x", $"${"x"}")
```

Because, Now, quotes in template literal's quotes of this spec is

- Should be written as `'${''x''}'`, `'${'x'}'` is invalid
- Should be written as `"${\"x\"}"`, `"${"x"}"` is invalid

but the spec of TypeScript is

- Should be written as `` `${`x`}` ``, escaping is not required

Is this good/bad idea?

aiya000 commented:
Also, I was forgeting removing supporting of

```vim
let x = 10
echo $'$x'  " 10
```

(https://github.com/vim/vim/pull/4634#issuecomment-557780193)

I'll remove this :+1:

aiya000 committed:
Remove supporting of variable expansion `$'$foo'` of template string from test

aiya000 committed:
Follow latest specs with document

aiya000 committed:
Remove supporting of variable expansion `$'$foo'` of template string from impl

aiya000 committed:
Make result of `get_template_string_tv()` as a concatenate style expression

brammool commented:
The question is:  Can $'---' first be handled like a '' string to find the end, and the ${} interpretation happens next, or do we need to handle the ${} items to find the end of the string?
In the first case quotes inside $'---' need to be doubled, just like with normal '---' strings.
In the second case the ${} item, and the expression inside it, needs to be evaluated to be able to find the end of the string.
I would think it's most important that the expression inside ${} can be written as a normal expression, without having to worry about another level of escaping.  The cost of finding the end of the expression is worth paying.

aiya000 commented:
Is it about interpretation orders and steps?

Thanks! I was not thinking about this.

In my opinion, it should be below.

```vim
let s:x = 10

function s:f() abort
  let s:x = 20
endfunction

function s:g() abort
  throw 'error'
endfunction

function s:h() abort
  let s:x = 30
endfunction
```

```vim
try
  echo $'${s:f()} ${s:g()} ${s:h()}'
catch
  echo s:x  " 20
  " Because these are evaluated the first to the end successively
  " if the template string is valid.
endtry
```

- - -

Also we should consider below case.

```vim
try
  " Missing closing quote
  echo $'${s:f()} ${s:h()}
catch
  echo s:x  " 30
endtry
```

This case cannot be happened on TypeScript and JavaScript,
    because that languages parses and raises the syntax error first,
    and the syntax error is not able to `catch`.

Compared with that languages, Vim script's syntax error is able to `catch`.

But I think below is better, because it has two benefits in my opinion.

1. The implementation can omit checking of the template string end's existance
1. The naturality. The syntax error occurs after the evaluation of `s:f()` and `s:h()`

aiya000 committed:
Fix memory leaks

aiya000 committed:
Fix warnings

aiya000 commented:
Maybe any issue & specs solved.

k-takata commented:
`:echo $"${'a"b'}"` shows `a` instead of `a"b`.

aiya000 commented:
Thanks!
I found a problem too X)

`echo $'${$UNDEFINED_ENV_VAR}'`  => segmentation fault

aiya000 committed:
Add tests for complex strings in strings

aiya000 committed:
Add a spec E450 for $UNDEFINED_ENV_VAR

k-takata commented:
How about this change? ~https://github.com/k-takata/vim/commit/8008fc3320dacde7ce4a9546034250dd3070e104~ https://github.com/k-takata/vim/commit/1961e54207f13800955a54d8966e88b0343a7886
Avoid using ~lambda~ the `STRINGIFY` string and avoid evaluating whole part of the template string. (Not tested well.)

aiya000 committed:
Fix $UNDEFINED_ENV_VAR evaluation

aiya000 commented:
@k-takata Thanks!
I'll read it :D

But it may conflict my curret commits.
Please allow me to take the commit manually!

k-takata commented:
I think that my change can supersede your latest 3 commits.
(I haven't checked whether it doesn't have any memory leaks yet, though.)

aiya000 commented:
Sorry, I want to use my commits... x(
But I don't want to abandon your commits.
I'll take my best to catch your effort!
Please allow me :pray:

In my opinion, below is important.
I'll take this! thanks!

- https://github.com/k-takata/vim/commit/1961e54207f13800955a54d8966e88b0343a7886#diff-bea881dfa9626bab7141337b0fcdb23eL1413
- https://github.com/k-takata/vim/commit/1961e54207f13800955a54d8966e88b0343a7886#diff-de9ab8906fef668ed8042046e0a1eb02R3903


And I couldn't understand below.
Why do you use `ga` instead of a part of `result`?
Please tell me :)

- https://github.com/k-takata/vim/commit/1961e54207f13800955a54d8966e88b0343a7886#diff-de9ab8906fef668ed8042046e0a1eb02R3702

Thanks a lot.

aiya000 committed:
Remove older specs from document

Thanks k-takata

aiya000 committed:
Improve tests about quotes in quotes

Thanks k-takata

aiya000 committed:
Fix wrong test

Thanks k-takata

k-takata commented:
> Why do you use `ga` instead of a part of `result`?

Consider parsing `$'foo${bar}baz'`.

* First accumulate the string into `ga`.
* When `${` is found, pass `ga` (= `'foo'`) to `eval1()` and concatenate the result into `result`. Then pass `bar` to `eval1()` and concatenate the result into `result`.
* Accumulate the rest of the string into `ga`.
* `ga` is now `'baz'`. Pass `ga` to `eval1()` and concatenate the result into `result`.
* Return `result`.

This way, you don't need to pass the result of `${bar}` to `eval1()` as you did.
(Your implementation evaluates `bar` with `eval1()`, then quotes it, then evaluates with `eval1()` again. It's not efficient, I think.)

aiya000 commented:
My implementation: (pseudo) `eval1('foo' .. eval1(bar) .. 'baz')`
Your implementation: (pseudo) `eval1('foo') .. eval1(bar) .. eval1('baz')`

Is this right?
The implementation should be this!

k-takata commented:
yes

aiya000 committed:
Eval current successively instead of whole eval

aiya000 committed:
Add tests for nested template string

k-takata commented on runtime/doc/eval.txt:
```diff
@@ -1383,6 +1383,60 @@ to be doubled.  These two commands are equivalent: >
 	if a =~ '\s*'
 
 
+template-string					*template-string* *E1000*
```
```suggestion
template-string				*template-string* *E451* *E452*
```

aiya000 replied:
👍

k-takata commented on runtime/doc/eval.txt:
```diff
@@ -1383,6 +1383,60 @@ to be doubled.  These two commands are equivalent: >
+	" Quotes in quotes
+	echo $'${'*'}'  " *
+	echo $"${"*"}"  " *
+	"" Meaning escaping is not needed
```
```suggestion
	" Meaning escaping is not needed
```

The meaning of this sentence is unclear to me. What does "Meaning" mean here?

aiya000 replied:
humm, maybe this sentence is unnecessary.
I'll remove below lines :)

```
	"" Escaping is not needed
	echo $'${''x''}'  " E199: ... (syntax error)
	echo $"${\"x\"}"  " E199: ... (syntax error)
```

k-takata commented on src/eval.c:
```diff
@@ -3665,6 +3674,367 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    current_result = eval_template_current(current, evaluate);
+    if (current_result == NULL)
+    {
```
```suggestion
```

`{` is not needed here.

aiya000 replied:
Thanks!
I'll do it.

k-takata commented on src/eval.c:
```diff
@@ -3665,6 +3674,367 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    if (current_result == NULL)
+    {
+	return FAIL;
+    }
```
```suggestion
```
`}` is not needed here.

k-takata commented on src/eval.c:
```diff
@@ -3665,6 +3674,367 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+     */
+    if (**arg == NUL)
+    {
+	ga_clear(&result);
```
Isn't `ga_clear(&current);` needed here?

aiya000 replied:
Just now I 'm really stuck by this memory leak.
Really thanks!

k-takata commented on src/eval.c:
```diff
@@ -3665,6 +3674,367 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    static char_u*
+stringify_expr(char_u *expr)
+{
+    const size_t    QUOTES_NUM = 2;  // opening and closing "
```
```suggestion
```
Unused.

aiya000 replied:
:+1:

k-takata commented on src/eval.c:
```diff
@@ -3665,6 +3674,367 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    **source = NUL;
+    expr = (char_u *) alloc(STRLEN(expr_head) + 1);
+    STRCPY(expr, expr_head);
+    **source = '}';
```
These 4 lines can be simpler:

```C
    expr = vim_strnsave(expr_head, (int)(*source - expr_head));
```

Or

```C
    return vim_strnsave(expr_head, (int)(*source - expr_head));
```

and remove the `expr` variable from this function.

aiya000 replied:
I'll use `    return vim_strnsave(expr_head, (int)(*source - expr_head));` !

k-takata commented on src/eval.c:
```diff
@@ -3665,6 +3674,367 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+	{
+	    ga_append(&current, (int) **arg);
+	    ++*arg;
+	    ga_append(&current, (int) **arg);
```
These 3 lines can be:

```C
	    ga_concatn(&current, *arg, 2);
	    ++*arg;
```

Or, if you don't want to use `ga_concatn()`, you can revert the changes to `misc2.c` and `misc2.pro`.

aiya000 replied:
I'll use `ga_concatn()`.
thanks!

aiya000 committed:
Update runtime/doc/eval.txt

Co-Authored-By: K.Takata <kentkt@csc.jp>

aiya000 committed:
Remove an unnecessary sentence from doc

aiya000 committed:
Refactor implementation

Thanks k-takata

aiya000 commented:
@k-takata Thanks!
~I might resolve all issues!~ new issue is found x(
Thanks a lot xD

I'm planning to use template string in my vimrc until Bram comes or new issue found.

k-takata commented on src/eval.c:
```diff
@@ -3665,6 +3673,350 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    {
+	ga_clear(&current);
+	ga_clear(&result);
+	clear_tv(&current_result);
```
Is this line needed? I think that `current_result` is not set, when `eval1()` returns `FAIL`.
Or, it might be safer to initialize `current_result` with `init_tv(&current_result);` at the top of this function.

aiya000 replied:
I don't know about `init_tv()`, but I think it should be used if it is safer than removing this line.
Should I use `init_tv()`?

k-takata commented on src/eval.c:
```diff
@@ -3665,6 +3673,350 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    {
+	result.vval.v_string = (char_u *) alloc(1);  // meaning success of this function
+	result.vval.v_string[0] = '\0';
+    }
```
This can be a bit simpler:

```suggestion
	result.vval.v_string = ALLOC_CLEAR_ONE(char_u);  // meaning success of this function
```

aiya000 replied:
thx!

k-takata commented on src/eval.c:
```diff
@@ -3665,6 +3673,350 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+/*
+ * Refreshes previous_quote with NUL
+ *	if the closing quote of
+ *	the opening quote (previous_quote) found on **source.
```
Can these lines be joined?
```suggestion
 *	if the closing quote of the opening quote (previous_quote) found on **source.
```

aiya000 replied:
thx!

aiya000 committed:
Update src/eval.c

Co-Authored-By: K.Takata <kentkt@csc.jp>

aiya000 committed:
Fix a memory leak

aiya000 committed:
Add a missing `init_tv()`

Before this

tmp.vim
```vim
if 0
  echo $"${10}"
endif
```

```shell-session
$ vim -u tmp.vim --noplugin
free(): invalid pointer
Vim: Caught deadly signal ABRT
Vim: Finished.
zsh: abort (core dumped)  ./src/vim -u tmp.vim --noplugin
```

aiya000 committed:
Refactor a little

Thanks mattn

aiya000 committed:
Refactor a little

Thanks k-takata

aiya000 committed:
Add tests

aiya000 committed:
Pass a test of lambda call

aiya000 committed:
Refactor a little

Thanks k-takata

k-takata commented on src/eval.c:
```diff
@@ -3665,6 +3673,337 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    }
+
+    if (current_result.vval.v_string == NULL)
+	current_result.vval.v_string = "";
```
```suggestion
	return FAIL;

    if (!evaluate)
	return OK;
```

aiya000 replied:
I'm really stupid. thanks!

aiya000 committed:
Update src/eval.c

Co-Authored-By: K.Takata <kentkt@csc.jp>

aiya000 commented:
I think any problems are solved! (again, again... x( )
I also git-rebased to master.

aiya000 commented:
And I made an independent branch https://github.com/aiya000/vim/commit/ff45f0cc26d8197ed2187de89390ed19c435833c .
This makes a syntax to show template strings by red.

![Screenshot_2020-03-04_14-57-48](https://user-images.githubusercontent.com/4897842/75850019-96821900-5e29-11ea-9a1e-5962cc395b7a.png)

I checked this can show test_template_string.vim correctly.

I'll send this to Dr. Chip after this is merged :+1:

aiya000 commented:
I hope to merge this into both Vim script and Vim9 script lol

aiya000 committed:
Add missing test_template_string.res for CI

aiya000 commented:
https://travis-ci.org/vim/vim/jobs/658088923?utm_medium=notification&utm_source=github_status is not this PR's fault, isn't it?

chrisbra commented:
looks like a flaky test. I restarted it.

k-takata commented:
(Off topic)
@chrisbra I thought I had a permission to restart Travis CI, but today I couldn't restart it.
Do you know anything?

chrisbra commented:
> @chrisbra I thought I had a permission to restart Travis CI, but today I couldn't restart it.
Do you know anything?

I had a quick look, but couldn't find where to configure membership and permissions. Let's discuss this offside of this channel here. Give me a mail please.

aiya000 commented:
Now, this can be merged :D

aiya000 commented:
I followed `origin/master` with `git-rebase` now :D

aiya000 committed:
Fix previous git-rebase origin/master

aiya000 commented:
I maybe fixed my fault :+1:

brammool commented:
Looking at this again (finally).
The help could use a few examples of useful application of this mechanism.  A constant inside the ${} isn't really useful.  Using a variable or function call would be good.
For evaluating the string is done with eval1().  Calling get_string_tv() or get_lit_string_tv() directly would be better.
The outer loop should increment by multi-byte characters instead of single bytes. MB_COPY_CHAR() can probably be used.

There is still a lot of copying going on.  For performance it won't be faster then using "aaa" .. expr .. "bbb".  For Vim9 script, where expressions are compiled, a separate implementation is required.  I have doubts if this is actually a good addition.

aiya000 commented:
> The help could use a few examples of useful application of this mechanism. A constant inside the ${} isn't really useful. Using a variable or function call would be good.
For evaluating the string is done with eval1(). Calling get_string_tv() or get_lit_string_tv() directly would be better.
The outer loop should increment by multi-byte characters instead of single bytes. MB_COPY_CHAR() can probably be used.

> There is still a lot of copying going on. For performance it won't be faster then using "aaa" .. expr .. "bbb".

I agree 👍️
I'll implement it :D

> For Vim9 script, where expressions are compiled, a separate implementation is required. I have doubts if this is actually a good addition.

Oh...
In Vim9 script, can this implementation not be used?
Can I improve this implementation for Vim9 script?

aiya000 commented:
Can I make new PR for Vim9 script if
this implementation cannot be used for Vim9 script? 

And, so, I believe this is useful for Vim(8) script writers :)

aiya000 commented:
Memo: Now

The concat style

```vim
function s:f() abort
  return 10
endfunction

let start = reltime()
for _ in range(100)
  let __ = 'aaa' .. s:f() .. 'bbb'
endfor
echo reltimestr(reltime(start))

" 0.000253
```

The template style

```vim
function s:f() abort
  return 10
endfunction

let start = reltime()
for _ in range(100)
  let __ = $'aaa${s:f()}bbb'
endfor
echo reltimestr(reltime(start))

" 0.000298
```

aiya000 committed:
Shorten  doc

aiya000 committed:
Eval using get_lit_string_tv() and get_string_tv()

aiya000 committed:
Increment each char bytes using MB_COPY_CHAR()

aiya000 commented:
@brammool 
I fixed tasks you gave me :)
As a result, the performance is improved!

- By above 'The concat style': almost 0.000253 ~ 0.000300 secs
- By above 'The template style': ~almost 0.000298 ~ 0.000360~ -> almost 0.000263 ~ 0.000320 secs
- - - - -

Another case, the printf style.
This is slower than the template style.

```vim
function s:f() abort
  return 10
endfunction

let start = reltime()
for _ in range(100)
  let __ = printf('aaa%sbbb', s:f())
endfor
echo reltimestr(reltime(start))

" 0.000300 ~ 0.000366
```

In my opinion,
    Usecases of the printf style and the template style are almost same.
    (Indeed, previously, I often used the printf style instead of the template style.)
But the template style has advantage.
Compared to `printf('a%s,b%s', F(), G())`, writing `$'a${F()}, b${G()}'` is hard to make bugs,
    and easy to understand.

- - - - -



The current implementation is more about `0.000015` secs than 'the concat style'. (106％ performance of the concat style.)

But in my opinion, it can't be helped, because Vim(8) script doesn't have compiling.

But I'll follow you, dear Bram :D
Please tell me more problems if these are existent!

aiya000 commented:
Also I `git-rebase`ed to origin/master :+1:

brammool commented:
> > For Vim9 script, where expressions are compiled, a separate
> > implementation is required. I have doubts if this is actually a good
> > addition.
> 
> Oh...
> In Vim9 script, can this implementation not be used?
> Can I improve this implementation for Vim9 script?

What your code is doing basically comes down to splitting the string at
occurences of ${}, evaluating the expression, and then concatenating
everything.  Thus:
	$'lead${ expr }tail'

Is evaluated as:
	'lead' .. (expr) .. 'tail'

For Vim9 script the "expr" part needs to be compiled.  The instructions
would become something like:
	PUSH 'lead'
	EVAL expr
	CONCAT
	PUSH 'tail'
	CONCAT

It's probably possible to share the parsing of the $'' string between
legacy Vim script and Vim9 script, and have two ways to produce output:
Either evaluate right away and return a typval, or produce instructions.

So it would be best if the implementation can be split into a part that
parses the string and returns a typval for the string part and a pointer
to the expression.  It would need to keep status of where it is in the
template, which could be just a pointer and/or a length.

-- 
Married is a three ring circus:
First comes the engagement ring.
Then comes the wedding ring.
Then comes the suffering.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

aiya000 commented:
Is meaning that we should get pairs of `(starting position of an element, ending position of the element)`?

```
"lead${expr}tail"
 ^     ^    ^
(0,3) (6,9) (11,14)
```

like

```
void getStartAndEndOnTemplate(char_u **result_starting_pos, char_u **result_ending_pos, char_u *template);

// for 'lead', expr, and 'tail'
// getting an array of { (0,3), (6,9), (11,14) }
for (...)
{
   char_u *s, *e;
   getStartAndEndOnTemplate(&s, &e, template);
}
```

?

aiya000 committed:
Split functions to support Vim9 script

aiya000 commented:
@brammool 

> So it would be best if the implementation can be split into a part that
> parses the string and returns a typval for the string part and a pointer
> to the expression.  It would need to keep status of where it is in the
> template, which could be just a pointer and/or a length.

I made a WIP commit with a type and a function.

- type `template_string_interval`: https://github.com/aiya000/vim/blob/template-string/src/eval.c#L65
- func `read_template_string_intervals`: https://github.com/aiya000/vim/blob/template-string/src/eval.c#L3476

Now we can get

```typescript
[
    { value: 'lead', is_expr: false },
    { value: 'expr', is_expr: true },
    { value: 'tail', is_expr: false },
]
```

from `"lead${expr}tail".

Is this helpful for Vim9 script?

aiya000 committed:
Add a test for postfix

aiya000 commented:
Now major implementing phase is finished!
The rest is only to fix memory leaks :D

aiya000 committed:
Remove unnecessary items

aiya000 committed:
Fix memory leaks

aiya000 committed:
Improve the performance

aiya000 commented:
Now this can be merged! :D

k-takata commented on src/eval.c:
```diff
@@ -3311,8 +3346,385 @@ eval_index(
+	if (ga_grow(result, 1) == FAIL)
+	{
+	    intervals_free(result);
+	    semsg(_("Internal error!"));
```
```suggestion
	    emsg(_(e_outofmem));
```
might be better?

aiya000 replied:
Thanks!

Same points was fixed!

aiya000 committed:
Update src/eval.c

Co-authored-by: K.Takata <kentkt@csc.jp>

Houl commented:
Error: thread too long

aiya000 commented:
@Houl Sorry, what do you mean? :)

aiya000 commented:
@brammool Hi, Bram :D
What do you think about latest implementation?

aiya000 commented:
@brammool Hi.
Should we include an implementation for vim9script into this PR?

brammool commented:
> @brammool Hi.
> Should we include an implementation for vim9script into this PR?

Eventually, yes.  Since it's an expression, which needs to be compiled.
But let's first see about the current implementation (sorry, didn't get
to looking at the most recent changes, bug repors keep me busy).

-- 
If Microsoft would build a car...
... You'd have to press the "Start" button to turn the engine off.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

aiya000 commented:
@brammool Thank you for your reply!
I'm so glad :D

I understood that you are in keeping busy :+1:

When you are no longer in busy, I would be grateful if you see the current implementation :D

Thank you!

Houl commented:
Is this actual syntax?

    " Quotes in quotes
    echo $'${'*'}'
    echo $"${"*"}"

If at all, I hope the syntax will be

    echo $'${''*''}'
    echo $"${\"*\"}"

otherwise it will be impossible to skip these new string literals using a single regexp.

aiya000 commented:
@Houl The first half is actual syntax :)
Maybe, A conversation of this is existent in this page 'Conversation'.

brammool commented:
> Is this actual syntax?
> 
>     " Quotes in quotes
>     echo $'${'*'}'
>     echo $"${"*"}"
> 
> If at all, I hope the syntax will be
> 
>     echo $'${''*''}'
>     echo $"${\"*\"}"
> 
> otherwise it will be impossible to skip these new string literals
> using a single regexp.

The expression inside the ${} will be a normal expression, no escaping
needed.  There also can be line breaks inside the ${}.  With Vim9 that
won't even require the leading backslash.

The trailing "}" is found by skipping over the expression, thus other
"}" can be inside the expression, so long as there is no syntax error.

This is very similar to Javascript/Typescript use of backtick strings.
It will be a challenge to recognize this with a regexp.

-- 
A real patriot is the fellow who gets a parking ticket and rejoices
that the system works.


 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

Houl commented:
The idea was to skip over `$'...'` or `$"..."` at once; and to omit the leading `$` and get an ordinary string literal.

Houl commented:
Wish: `exprend({str}[, {start}])`, a function to skip over a Vim expression.  It should be something that's already there, just not exposed yet.

aiya000 commented:
> Wish: exprend({str}[, {start}]), a function to skip over a Vim expression. It should be something that's already there, just not exposed yet.

Is this meaning the C function?
If this is meaning the C function, just `get_template_string_tv()` is useful.

austintraver commented:
@aiya000 I've been following this since [your original post in June of 2019](https://github.com/vim/vim/issues/4491) and I'm so impressed with your patience and commitment to getting this integrated into `vi`. I'm a huge fan of your proposal, you have my support.

aiya000 commented:
@austintraver 
Waw! I'm so glad!

This PR was so hard for me, because this is the first contribution of Vim.

Thank you :D
Thank you!
